core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1963 @@
+
+    /// Returns the edge left to `handle`.
+    fn left_edge(&self, handle: KVHandle) -> KVHandle {
+        KVHandle::new(handle.node(), handle.idx())
+    }
+
+    /// Returns the edge right to `handle`.
+    fn right_edge(&self, handle: KVHandle) -> KVHandle {
+        KVHandle::new(handle.node(), handle.idx() + 1)
+    }
+
+    /// Returns a handle to the key/value pair left of `handle`.
+    ///
+    /// If `handle` already points to the first element in the node
+    /// an `Err(handle)` is returned. The handle contained in `Err(handle)`
+    /// is the one supplied as the `handle` argument to this function.
+    fn left_kv(&self, handle: KVHandle) -> Option<KVHandle> {
+        if handle.idx() > 0 {
+            Some(KVHandle::new(handle.node(), handle.idx() - 1))
+        } else {
+            None
+        }
+    }
+
+    /// Returns a handle to the key/value pair right of `handle`.
+    ///
+    /// If `handle` already points to the last element in the node an
+    /// `Err(handle)` is returned. The handle contained in `Err(handle)`
+    /// is the one supplied as the `handle` argument to this function.
+    fn right_kv(&self, handle: KVHandle) -> Option<KVHandle> {
+        let node = self
+            .get_node(handle.node())
+            .expect("node to descend from must exist");
+        if handle.idx() < node.len() {
+            Some(handle)
+        } else {
+            None
+        }
+    }
+
+    /// Removes the key/value pair pointed to by `handle`.
+    ///
+    /// If through this removal an underfull node was created, appropriate strategies
+    /// will be deployed (`.handle_underfull_node`).
+    ///
+    /// Returns the removed value.
+    fn remove_kv(&mut self, handle: KVHandle) -> V {
+        debug_assert!(self.header.len > 0);
+        self.header.len -= 1;
+
+        let handle_type = self.get_handle_type(handle.node());
+        let (small_leaf, old_pair_ptr, mut new_len) = match handle_type {
+            Leaf => self.extract_handle(handle),
+            Internal => {
+                let child = self
+                    .right_child(handle)
+                    .expect("every internal node has children; qed");
+                let first_leaf = self.first_leaf_edge(child);
+
+                let to_remove = self.right_kv(first_leaf).expect("right_kv must exist");
+                let (hole, pair_ptr, nl) = self.extract_handle(to_remove);
+
+                let node = self.get_node_mut(handle.node()).expect("node must exist");
+                let old_pair_ptr = node
+                    .set_pair(handle.idx(), Some(pair_ptr))
+                    .expect("handle must be valid");
+                (hole, old_pair_ptr, nl)
+            }
+        };
+
+        let mut handle = small_leaf.node();
+        while new_len < CAPACITY / 2 {
+            match self.handle_underfull_node(handle) {
+                UnderflowResult::AtRoot => break,
+                UnderflowResult::EmptyParent(_) => unreachable!(),
```

Comment:
Please add a comment in the `unreachable!` providing a rational that states exactly why it is unreachable.